crates/bevy_ecs/src/schedule/graph/dag.rs

```diff
@@ -0,0 +1,995 @@
+                (Err(lhs_key), Ok(rhs)) => {
+                    // Edge from a key node to a value node, expand to all values in the key's group
+                    for &lhs in self.get(&lhs_key).into_iter().flatten() {
+                        flattened.add_edge(lhs, rhs);
+                    }
+                }
+                (Ok(lhs), Err(rhs_key)) => {
+                    // Edge from a value node to a key node, expand to all values in the key's group
+                    for &rhs in self.get(&rhs_key).into_iter().flatten() {
+                        flattened.add_edge(lhs, rhs);
+                    }
+                }
+                (Err(lhs_key), Err(rhs_key)) => {
+                    // Edge between two key nodes, expand to all combinations of their value nodes
+                    for &lhs in self.get(&lhs_key).into_iter().flatten() {
+                        for &rhs in self.get(&rhs_key).into_iter().flatten() {
+                            flattened.add_edge(lhs, rhs);
+                        }
+                    }
+                }
+            }
+        }
+
+        flattened
+    }
+}
+
+impl<K, V, S> Deref for DagGroups<K, V, S> {
+    type Target = HashMap<K, HashSet<V, S>, S>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<K, V, S> DerefMut for DagGroups<K, V, S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<K, V, S> Default for DagGroups<K, V, S>
+where
+    S: BuildHasher + Default,
+{
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<K: Debug, V: Debug, S> Debug for DagGroups<K, V, S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("DagGroups").field(&self.0).finish()
+    }
+}
+
+/// Error indicating that the graph has redundant edges.
+#[derive(Error, Debug)]
+#[error("DAG has redundant edges: {0:?}")]
+pub struct DagRedundancyError<N: GraphNodeId, S = FixedHasher>(pub HashSet<(N, N), S>);
+
+/// Error indicating that two graphs both have a dependency between the same nodes.
+#[derive(Error, Debug)]
+#[error("DAG has a cross-dependency between nodes {0:?} and {1:?}")]
+pub struct DagCrossDependencyError<N>(pub N, pub N);
+
+/// Error indicating that the graph has overlapping groups between two keys.
+#[derive(Error, Debug)]
+#[error("DAG has overlapping groups between keys {0:?} and {1:?}")]
+pub struct DagOverlappingGroupError<K>(pub K, pub K);
+
+#[cfg(test)]
+mod tests {
+    use core::ops::DerefMut;
+
+    use crate::schedule::graph::{index, Dag, Direction, GraphNodeId, UnGraph};
+
+    #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    struct TestNode(u32);
+
+    impl GraphNodeId for TestNode {
+        type Adjacent = (TestNode, Direction);
+        type Edge = (TestNode, TestNode);
+
+        fn kind(&self) -> &'static str {
+            "test node"
+        }
+    }
+
+    #[test]
+    fn test_dirty_on_deref_mut() {
```

Comment:
seems to be testing more than deref_mut.
I'd also remove `test_` from all the test names. It's redundant since they're all just tests.
```suggestion
    fn mark_dag_dirty() {
```